MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -255,6 +262,284 @@ SetUefiImageMemoryAttributes (
+                                 );
+      if (SpecialRegionEntry->SpecialRegion.Start > NextSpecialRegionEntry->SpecialRegion.Start) {
+        // Temp = Current
+        TempSpecialRegion.Start         = SpecialRegionEntry->SpecialRegion.Start;
+        TempSpecialRegion.Length        = SpecialRegionEntry->SpecialRegion.Length;
+        TempSpecialRegion.EfiAttributes = SpecialRegionEntry->SpecialRegion.EfiAttributes;
+
+        // Current = Next
+        SpecialRegionEntry->SpecialRegion.Start         = NextSpecialRegionEntry->SpecialRegion.Start;
+        SpecialRegionEntry->SpecialRegion.Length        = NextSpecialRegionEntry->SpecialRegion.Length;
+        SpecialRegionEntry->SpecialRegion.EfiAttributes = NextSpecialRegionEntry->SpecialRegion.EfiAttributes;
+
+        // Next = Temp
+        NextSpecialRegionEntry->SpecialRegion.Start         = TempSpecialRegion.Start;
+        NextSpecialRegionEntry->SpecialRegion.Length        = TempSpecialRegion.Length;
+        NextSpecialRegionEntry->SpecialRegion.EfiAttributes = TempSpecialRegion.EfiAttributes;
+      }
+
+      NextSpecialRegionEntryLink = NextSpecialRegionEntryLink->ForwardLink;
+    }
+
+    SpecialRegionEntryLink     = SpecialRegionEntryLink->ForwardLink;
+    NextSpecialRegionEntryLink = SpecialRegionEntryLink->ForwardLink;
+  }
+}
+
+/**
+  Copy the HOB MEMORY_PROTECTION_SPECIAL_REGION entries into a local list
+
+  @retval EFI_SUCCESS           HOB Entries successfully copied
+  @retval EFI_OUT_OF_RESOURCES  Failed to allocate
+**/
+STATIC
+EFI_STATUS
+CollectSpecialRegionHobs (
+  VOID
+  )
+{
+  EFI_HOB_GUID_TYPE                            *GuidHob          = NULL;
+  MEMORY_PROTECTION_SPECIAL_REGION             *HobSpecialRegion = NULL;
+  MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY  *NewSpecialRegion = NULL;
+
+  DEBUG ((
+    DEBUG_INFO,
+    "%a - Enter...\n",
+    __FUNCTION__
+    ));
+
+  GuidHob = GetFirstGuidHob (&gMemoryProtectionSpecialRegionHobGuid);
+
+  while (GuidHob != NULL) {
+    DEBUG ((DEBUG_INFO, "%a - 1\n", __FUNCTION__));
+    HobSpecialRegion = (MEMORY_PROTECTION_SPECIAL_REGION *)GET_GUID_HOB_DATA (GuidHob);
+    NewSpecialRegion = AllocateCopyPool (sizeof (MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY), HobSpecialRegion);
+
+    if (NewSpecialRegion == NULL) {
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    CopyMem (&NewSpecialRegion->SpecialRegion, HobSpecialRegion, sizeof (MEMORY_PROTECTION_SPECIAL_REGION));
+    NewSpecialRegion->Signature = MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY_SIGNATURE;
+    InsertHeadList (&mSpecialMemoryRegionsPrivate.SpecialRegionList, &NewSpecialRegion->Link);
+    mSpecialMemoryRegionsPrivate.Count++;
+
+    GuidHob = GetNextGuidHob (&gMemoryProtectionSpecialRegionHobGuid, GET_NEXT_HOB (GuidHob));
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Create a sorted array of MEMORY_PROTECTION_SPECIAL_REGION structs describing
+  all memory protection special regions.

```

Comment:
can you add a comment about the allocated memory needing to be freed by caller?

Reply:
Added a comment in the function header stating that the caller must free the allocated array.